### ISSUE TITLE
Conversion Hangs Creating Multiple Files by Chapter

### ISSUE THREAD

**Issue by BDG61** [audiamus/AaxAudioConverter#186]
Hello, First off, thanks for creating this fantastic program! 

Program works fine creating "Single file by AAX file" but hangs/freezes creating "Multiple files, by chapter". No error message displayed. AAX Audio Converter dialog box just freezes after completing several steps and becomes unresponsive to Abort or Close X clicks. Must use Task Manager to End Task in order to close program. Tried several book titles. All hang during conversion by chapters. Source Audible files were downloaded successfully using Book Lib Connect. Default settings being used during conversion, with exception of requesting creation of Multiple files by chapter. 

Please advise what I must do to have files convert smoothly by chapter. Thanks.

Level 3 log file attached. 
[AaxAudioConverter_2023-08-01_001.log](https://github.com/audiamus/AaxAudioConverter/files/12231397/AaxAudioConverter_2023-08-01_001.log)

**Comment by audiamus** [comment]
I can't yet see any hint from the log file. Some chapters appear to be transcoded normally and complete their task. But this all runs in parallel threads. For narrowing down the problem, can you disable parallel processing? In the installation folder, you'll find an XML file `AaxAudioConverter.exe.config`. Change the `value` in 
```
      <setting name="NonParallel" serializeAs="String">
        <value>False</value>
      </setting>
```
to `True` for this. (It will slow down the conversion, of course.)

And also increase the log level to 4, please.

**Comment by BDG61** [comment]
Success! Changed NonParallel value to "True" and ran with -Log=4. 135MB .m4b file converted into mp3 chapters in about 4min. Ideally, I'd like to be able to convert using multiple threads. Device and OS details below along with level 4 log file. 

Device name	Lenovo X280 
Processor	Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz   1.99 GHz
Installed RAM	8.00 GB (7.85 GB usable)
Product ID	00330-51288-24181-AAOEM
System type	64-bit operating system, x64-based processor
Pen and touch	Touch support with 10 touch points

Edition	Windows 10 Pro
Version	22H2
OS build	19045.3208
Experience	Windows Feature Experience Pack 1000.19041.1000.0

Log file attached.
[AaxAudioConverter_2023-08-01_002.log](https://github.com/audiamus/AaxAudioConverter/files/12233935/AaxAudioConverter_2023-08-01_002.log)

**Comment by audiamus** [comment]
Weird. Looks like some sort of system resource problem. This i7 has many CPU cores, I presume. But main memory could be a bit on the low side for exploiting all the CPU cores' potential - which AAX Audio Converter does.  

"NonParallel" still runs with worker threads (as you can see from the tread ID in the 3rd column of the log), but it throttles them down to a trickle. All the audio processing is done in FFmpeg, which is launched as a sub-process for each audio handling worker thread. Log level 4 reports FFmpeg setup and console output. 

AAX Audio Converter comes with two different builds of FFmpeg. One is an older build, among the last they did as 32bit, but it is usually a bit faster than later 64bit builds, so it's the preferred choice. FFmpeg may allocate lots of memory and your system may run out of it. Therefore I suggest to enforce the much newer 64bit build of FFmpeg for the next test. It might tackle memory consumption more conservatively. You do that by setting the value of `FFmpeg64bitHours` to `1` in the config file. Reset `NonParallel` back to `False`.

**Comment by BDG61** [comment]
Given your comment about possibly being limited by memory, I reset NonParallel to False, closed multiple browser windows I had open and reran same source file trying to create multiple chapters. Seemed to be running threads. Program hung as before with no error message - just unresponsive. 

Then changed FFmpeg64bitHours to 1 and reran with -Log=4 (NonParallel still set to False). Program hung as before. Log file attached below.

[AaxAudioConverter_2023-08-02_001.log](https://github.com/audiamus/AaxAudioConverter/files/12242677/AaxAudioConverter_2023-08-02_001.log)

**Comment by audiamus** [comment]
I'm afraid `FFmpeg64bitHours=1` didn't have the desired effect, it was still using the 32bit build. You will have to manipulate the executable names instead. In the AAX Audio Converter folder, where you also found the config file. rename `ffmpeg.exe` to `ffmpeg32.exe`. Then copy `ffmpeg64.exe` to `ffmpeg.exe`. Now both ffmpeg executables are 64bit. Run the previous test again.

Nonetheless, I think your system is at its limit when running in parallel mode. In indicator for this is the transcoding speed of the FFmpeg processes. In non-parallel mode speed is higher than 70x. In parallel mode, it drops down to 10 - 15x. That's not the purpose of parallel processing. It should utilize the CPU cores and run at full speed on each of them. Unfortunately there is no setting in AAX Audio Converter to throttle parallelization, it's either on or off.

**Comment by BDG61** [comment]
I renamed ffmpeg.exe to ffmpeg32.exe and then renamed ffmpeg64.exe to ffmpeg.exe and reran same source file. Program still froze but seemed less insulted. This time I was able to move its frame around the screen by clicking and dragging on the title bar and although the Abort button would not respond, I was able to close the program using the red X in the top right corner (as opposed to having to use Task Manager). Log file attached.
  
[AaxAudioConverter_2023-08-02_002.log](https://github.com/audiamus/AaxAudioConverter/files/12243779/AaxAudioConverter_2023-08-02_002.log)

**Comment by audiamus** [comment]
The newer FFmpeg now but not much difference otherwise. 

Question: Do you absolutely need .mp3? Or would .m4b also do? In any case, you can try for test purposes to "convert" to .m4b, still in split by chapter mode. That would run FFmpeg as a copy operation, with no transcoding involved. The overall behavior might be different .

**Comment by BDG61** [comment]
I tried running as before except to output multiple chapters as .m4b. Program froze and needed Task Manager to close it. Log below.
[AaxAudioConverter_2023-08-02_003.log](https://github.com/audiamus/AaxAudioConverter/files/12244302/AaxAudioConverter_2023-08-02_003.log)

In answer to your question, I'm after .mp3 files to play Audible books on an old Blackberry Z30 running BB OS10 (and a version of Android 4.3) since Android 4.3 can't run the current Audible app and the old app made for 4.3 seems unwelcome at Audible servers (I can login, but not see my library). Now, with .mp3 book chapters, I can use the native OS10 music player or an audio book player. It's also nice to have a backup of my audio books. Non-parallel processing will just take a bit longer to run. I suppose I could also try running conversion on another PC with more memory. 

I gather best course of action now on my X280 is to: 1) return NonParallel parameter to True in config file, 2) rename ffmpeg.exe back to ffmpeg64.exe and 3) rename ffmpeg32.exe back to ffmpeg.exe?

**Comment by FooDeas** [comment]
I have the same problem here. I have more then 20GB memory free during the conversion.
Setting running single threaded "solves" the problem. @audiamus May I try something else to help solving the issue?
